### ISSUE TITLE
[CodeGen] Add Support for NativeField and NativeMethod

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#410]
The code generator should emit code for the NativeField and NativeMethod expressions.

NB: *Not* via a reflective call through `invoke`, but through a real, direct call.

**Comment by magnus-madsen** [comment]
This one should be fairly straightforward, but test cases should be written, since it is a new feature.

**Comment by FlaminGuy** [comment]
This is a little bit tricky because the way things are implemented currently, when we search for a method or a constructor we filter available methods by their name and number of parameters and if there is more than one method we fail. 
For example `unsafe native method java.lang.Math.abs(-1)` fails because the single parameter of `abs` might have different types. 
Now this can be fixed by passing parameters to `namer` then we have to use something similar to  `descriptor` in `Codegen.scala` but I'm not quite sure how we can get `interfaces` map in `namer`.
Let me know if you know how to quickly have that map, otherwise I'll continue investigating it.

**Comment by magnus-madsen** [comment]
You are right. Good catch :) In general we can't call any java constructor/method.

The resolution procedure only works in the case where the constructor/method name and the number of arguments uniquely determine the constructor/method to be invoked.

Unfortunately we cannot support Java-style overloading of the same method name, since the Flix type system is incompatible with the Java type system. 

For now the goal is just to be able to call constructors/methods where there is no ambiguity. 

In other words, you only have to worry about CodeGen and the tests you write have to call unambiguous constructors/methods.

**Comment by FlaminGuy** [comment]
I believe there is a way around this, maybe not fully supporting everything but I think we can do a way better job that doesn't take that long.
We can talk about this tomorrow.

**Comment by magnus-madsen** [comment]
The fundamental problem is that the type system of Flix is based on unification whereas the type system of Java is based on sub-typing. We don't want to add sub-typing to Flix, so in general we won't be able to type check code that interops with Java. This then leads to problems where Java requires type knowledge to select the appropriate method, and this type knowledge might not be available to us.

**Comment by mhyee** [comment]
Could the user provide (Java) type information by writing annotations in the Flix code?

**Comment by magnus-madsen** [comment]
Yes, it could work for primitives: `native method java.lang.Math.abs(x oftype Float)`, but for classes we would still need to deal with them somehow, or you would have to write:

 `native method java.lang.Foo.bar(x oftype java.lang.xyz, y, oftype java.lang.xyz)`

which would suck.

**Comment by magnus-madsen** [comment]
The point about the current system is that it would allow relatively clean interop with Java code that doesn't require overloading etc. That means you could be able to write some nice interop layer between Java and Flix. For example, we could define our own `File` class which does not make use of overloading, if so desired.

**Comment by FlaminGuy** [comment]
I have the code for calling a native method when there is only one choice of method/constructor based on number of parameters and I'll push it shortly, but it's so limited that I have hard time creating tests for it. 
Most of the time, we only can call classes with zero parameter constructor and most of generic methods are overloaded anyway. Honestly I don't see any use for this feature.       
My recommendation is to do one of the following:
1- delay the choice of which method to pick to when we either have to invoke it or generate code for it. Most of the times we have the types for parameters at that stage so we can pick the right method. In case of conversions, using `Class[_]`, we can find out the hierarchy of each type too and we can use them to find the right method. I know even then there is edge cases but by doing this, we go from 10% functionality to +90% without adding any new edge cases. 
2- Type the params like you suggested, maybe a better syntax is to do `x: java.lang.xyz`?
3- Completely scrap this native call idea and use the `File` class idea that Magnus has. When there is nothing to call why make it available?

**Comment by magnus-madsen** [comment]
> I have the code for calling a native method when there is only one choice of method/constructor based 
> on number of parameters and I'll push it shortly, but it's so limited that I have hard time creating tests for it.
> Most of the time, we only can call classes with zero parameter constructor and most of generic methods > are overloaded anyway. Honestly I don't see any use for this feature.

It might be limited, but it is better than nothing :) If you are aware of the restrictions it is easy to write a bridge layer for any functionality you have in mind. We can use this to significantly extend the standard library available to Flix programs.

> My recommendation is to do one of the following:
> 1- delay the choice of which method to pick to when we either have to invoke it or generate code for it. Most of the times we have the types for parameters at that stage so we can pick the right method. In case of conversions, using Class[_], we can find out the hierarchy of each type too and we can use them to find the right method. I know even then there is edge cases but by doing this, we go from 10% functionality to +90% without adding any new edge cases.

I think it is a hard problem. We can still think about it and see if we can come up with some satisfactory solution. To give you an idea of some of the challenges, just think about stuff like:

```java
int foo(int x);
float foo(float x);
int bar(int x);
float bar(float x);
```

and then a call like `bar(foo(x))` where the type of `x` is not known. 

I think down the road, when we get type classes, we might be able to do something a bit better.

> 2- Type the params like you suggested, maybe a better syntax is to do x: java.lang.xyz?

If you are willing to fully annotate the type parameters of the function when you call it then we can resolve it, but it will be verbose, and then question is also what the syntax should look like.

> 3- Completely scrap this native call idea and use the File class idea that Magnus has. When there is nothing to call why make it available?

Oh, but the idea is that we write a java class, say `MyFile` or `FlixFile` and we take all these limitations into account, and then we expose that to the Flix programs.

**Comment by FlaminGuy** [comment]
Makes sense.

**Comment by magnus-madsen** [comment]
Implemented.
